Move most docs to sphinx

ddn0 opened the pull request:
Put documentation in a single website with Sphinx rather than having
information spread across markdown files that are implicitly organized
by the repo directory structure.

witchel commented on libtsuba/src/azure.cpp:
```diff
@@ -0,0 +1,161 @@
+
+  std::unique_ptr<az::blob_client> client = std::move(client_res.value());
+  auto future = client->download_blob_to_buffer(container, blob, start, size,
+                                                result_buf, 1);
```
Where is download_blob_to_buffer defined?

tylershunt replied:
Here: https://github.com/Azure/azure-storage-cpplite/blob/11e1f98b021446ef340f4886796899a6eb1ad9a5/src/blob/blob_client.cpp#L109

witchel replied:
Thanks.  What is weird is that searching for this function here fails.
https://azure.github.io/azure-storage-cpp/

I do appreciate how the azure code seems to have fewer annoying wrappers than s3.

tylershunt replied:
Yeah the non-lite sdk has some pretty heave dependencies on cassablanca and it seems like they only provide more features (e.g., azure also has a file storage service). The lite version seemed more manageable esp if our goal is just to get some azure compatibility at first.

ddn0 commented:
An alternative to including the azure client always would be to conditionally build it like we do with mongodb in tools/graph-convert/CMakeLists.txt. It seems a little strange to require both the AWS S3 library and the Azure Blob library when users usually have to interact with one or the other.

arthurp commented:
> None of the azure storage SDKs are available though conda forge so I
> opted for CMake FetchContent instead. Download happens at configure
> time.

That's not likely to work since unless the Azure package is statically linked against libtsuba. Is it?

As for the build failure, it looks like there is a missing uuid library in the cmake configure. It's not totally clear to me where uuid is required, but adding `libuuid` as a build dependency in `conda_recipe/meta.yaml` might fix it.

arthurp commented:
> An alternative to including the azure client always would be to conditionally build it like we do with mongodb in tools/graph-convert/CMakeLists.txt. It seems a little strange to require both the AWS S3 library and the Azure Blob library when users usually have to interact with one or the other.

Ideally we would build the mongodb, Azure, and S3 support as separate `.so`s, just like they already have separate headers. That way we could package them separately (galois-s3, galois-mongodb, etc) and have the correct dependencies on the feature packages without weighing down the main package. Without separate packaging of these features we are pretty restricted by the dependencies of the prebuilt packages.

This is kind of a separate issue, but I thought it's worth bringing up.

tylershunt commented:
> That's not likely to work since unless the Azure package is statically linked against libtsuba. Is it?

Yes The way I've configured it the Azure package should always be statically linked with `libtsuba`. It seemed like the fastest way to converge.

tylershunt commented:
> It's not totally clear to me where uuid is required, but adding `libuuid` as a build dependency in `conda_recipe/meta.yaml` might fix it.

Ah this is a dependency of azure, I'll figure out what to do about it.

tylershunt commented:
> An alternative to including the azure client always would be to conditionally build it like we do with mongodb in tools/graph-convert/CMakeLists.txt. It seems a little strange to require both the AWS S3 library and the Azure Blob library when users usually have to interact with one or the other.

Sounds good, I'll add another config flag.

tylershunt commented on libtsuba/src/azure.cpp:
```diff
@@ -0,0 +1,161 @@
+}
+
+galois::Result<std::unique_ptr<az::blob_client>> GetClient() {
+  std::string account_name;
```
Actually in `s3.cpp` a new client is constructed in very much the same way everytime we call `GetClient`. I think the `shared_ptr` in s3's version is actually unnecessary, we should probably change it to a unique pointer (or work out if having one global client is better/matters).

arthurp commented:
> Ah this is a dependency of azure, I'll figure out what to do about it.

See https://github.com/KatanaGraph/katana/blob/master/CONTRIBUTING.md#adding-new-external-dependencies The package is `libuuid` I'm pretty darn sure.

tylershunt commented:
My last commit let's the builder specify which backend is built at config time. Let me know what you think!

ddn0 commented on libtsuba/CMakeLists.txt:
```diff
@@ -56,14 +53,56 @@ else()
+endif()
+
+if (TSUBA_BACKEND_local)
+  target_sources(tsuba PRIVATE src/LocalStorage.cpp)
```
sweet I didn't realize that `target_sources` appends

ddn0 commented on libtsuba/CMakeLists.txt:
```diff
@@ -56,14 +53,56 @@ else()
+  message(FATAL_ERROR "Must provide a storage backend")
+endif()
+
+foreach(BACKEND ${GALOIS_STORAGE_BACKEND})
```
some preference for local variables to just be snake_case, e.g. `backend`, `lc_backend`, `tsuba_backend_azure`. (not that cmake has any great variable hiding capabilities).

edit: I see that `TSUBA_BACKEND_s3` is used in tests/CMakeLists.txt, so maybe `TSUBA_BACKEND_S3` (capital S) would also be ok. Too bad the benchmarker can't run with various backends...

tylershunt replied:
My plan is to adapt it to be backend agnostic when I go through the exercise of  collecting some azure numbers.

tylershunt commented:
Was having build trouble with conda rather than sink a bunch of time into it now I decided to fix the conda version to what it was before (only s3 + local fs) and leave conda + azure for another day. Documented in issue #416.

tylershunt commented on libtsuba/src/GlobalState.cpp:
```diff
@@ -3,26 +3,49 @@
+#endif
+#ifdef GALOIS_HAVE_LOCAL_BACKEND
+    &tsuba::local_storage_allocator,
+#endif
```
This is the best I could come up with to get the static link build to work.

ddn0 replied:
seems reasonable to me.

tylershunt commented on libtsuba/include/tsuba/azure_internal.h:
```diff
@@ -0,0 +1,37 @@
+namespace tsuba::internal {
+
+class AzureAsyncWork : public FileAsyncWork {
+  std::future<galois::Result<void>> future_;
```
I ended up using `std::async` and `std::future` to implement `FileAsyncWork` for azure because it returns futures anyway. I think we might want to consider `std::future` for `s3` as well then we could drop `AsyncWork`. We can use `std::async` to let the compiler schedule our code on background threads or get more control with something like `std::packaged_task`.

witchel replied:
I keep reading about C++ threading and every session ends with me being confused.  Do I care if the std::future destructor blocks?  What is it blocking for?  But I like this summary:

Use std::async if you want some things done and don't really care when they're done, and std::packaged_task if you want to wrap up things in order to move them to other threads or call them later.

But I remember looking to define a thread pool that would just execute tasks from a work queue, and this functionality was "planned for C++20."  Maybe I missed something.

As far as FileAsyncWork, I used a stack of functions when I first wrote it, but it would be easy to have each function enqueue the next.

tylershunt replied:
I think `std::async(function());` is essentially `std::thread(std::packaged_task(function());`. The benefit of `FileAsyncWork`, as I see it, is that the caller can enforce some control over task priority. It would be easy/simpler to write: `future = std::async([](){ func1(); func2(); return func3();});` for example and just have the user call `future.get()` (again at the cost of some control, but it boils down to control of which network events are waited on first in this case).

Futures represent some completed execution so if you never call `get` someone has to block to make sure all of the expected code was run, so the destructor does.

tylershunt replied:
Anyway, the pertinent thing for this checkin is that I didn't think that level of control was necessary/attractive for azure since it's going to get less attention.

witchel commented on libtsuba/include/tsuba/s3_internal.h:
```diff
@@ -24,7 +26,8 @@ class S3AsyncWork : public FileAsyncWork {
-  std::stack<galois::Result<void> (*)(S3AsyncWork& s3aw)> func_stack_{};
+  std::stack<std::function<galois::Result<void>(S3AsyncWork& s3aw)>>
+      func_stack_{};
   std::string token_{};
```
Ah, I'm so proud for having mastered C-style function pointers, I forget to stop using them in C++.

tylershunt replied:
They are technically lighter weight! So you weren't wrong, I just needed to be able to pass lambdas that capture state as well.

tylershunt commented on libtsuba/include/tsuba/FileAsyncWork.h:
```diff
@@ -1,18 +1,12 @@
-  std::unordered_set<std::string> listing_{};
-
-protected:
 public:
```
Changed this `set` to be a `*` parameter to `List`

tylershunt commented on libtsuba/include/tsuba/s3_internal.h:
```diff
@@ -24,7 +26,8 @@ class S3AsyncWork : public FileAsyncWork {
 
-  std::stack<galois::Result<void> (*)(S3AsyncWork& s3aw)> func_stack_{};
+  std::stack<std::function<galois::Result<void>(S3AsyncWork& s3aw)>>
+      func_stack_{};
```
The function object here lets me use lambdas that capture state as well as raw function pointers.

tylershunt commented:
My apologies to the reviewers, when I started I didn't intend for this to be a *checks notes* 800 line merge.

tylershunt commented:
Done playing with it now. Tested using @witchel's generate graphml->convert to azure/s3 -> run `tsuba_fault` a bunch -> run `tsuba_gc -r 1`

arthurp commented on README.md:
````diff
@@ -4,59 +4,65 @@ Overview
+Then, install Katana
+
+```bash
+conda install -c katanagraph katana katana-python
````
This will not currently work because we do not have any "full" releases. The only packages in conda are in `katanagraph/label/dev`. So we should include getting the development builds here.

ddn0 replied:
Updated to the following:

=======================

Then, install Katana

```bash
conda install -c katanagraph/label/dev katana katana-python
```

Currently, only development versions are available through Conda. These builds
reflect the current state of the `master` branch. Releases of stable versions
will be available in the future.

arthurp commented on docs/contributing/building.rst:
```diff
@@ -0,0 +1,221 @@
+   cmake -DCMAKE_TOOLCHAIN_FILE=conan_paths.cmake \
+     -DLLVM_DIR="$(llvm-config-X --cmakedir)" ..
+
+Adding New External Dependencies
```
I liked the "eternal" pun!

ddn0 replied:
Reverted.

arthurp commented on scripts/setup_ubuntu.sh:
```diff
@@ -22,7 +22,6 @@ VERSION=$(lsb_release --release --short | cut -d . -f 1) # numerical major versi
 SETUP_TOOLCHAIN_VARIANTS=${SETUP_TOOLCHAIN_VARIANTS:-yes}
```
I support these changes, but they are out of place in the PR. Up to you.

ddn0 replied:
You mean the addition of `libnuma-dev`? Let me put this as a separate commit. The reason it came up was that I was following our build instructions in `Python.md` and it mentioned `libnuma-dev` but I switched the instructions to reference `scripts/setup_ubuntu.sh` for simplicity, so it made sense to at least add `libnuma-dev` to `setup_ubuntu.sh`.

arthurp replied:
Ah. I see. I actually cannot find a reference to this script. I do see `scripts/setup_dev_ubuntu.sh`, but I would rather not use that for conda builds. It installs a HUGE amount of extra stuff that is not needed for the conda build and does so at the system level. It also initializes conan which you should NEVER use with conda, so I would rather not encourage people to use it.

All that said, not having numa setup correctly doesn't prevent a correct build. It just prevents NUMA from working correctly at runtime.

arthurp replied:
I suspect the Python build instructions could use some tweaking anyway. So I can do that in a separate PR. So continue as you see fit.

ddn0 replied:
`setup_ubuntu.sh` is referenced from `setup_dev_ubuntu.sh`.

```
katana/scripts/setup_dev_ubuntu.sh
28:sudo bash -x "${REPO_ROOT}/scripts/setup_ubuntu.sh" --no-setup-toolchain-variants
```

I'm going to leave libnuma in `setup_ubuntu.sh` because I deleted from Python.md. But yeah, feel free to streamline the `setup_dev_ubuntu.sh` for conda versus conan.

arthurp reviewed (approved):
This makes me *SO VERY HAPPY!!!!!*

The first comment about dev packages is the only one that needs to be addressed really.

ddn0 committed:
Install libnuma-dev

Update scripts/setup_ubuntu.sh to match the instructions in Python.md.

ddn0 committed:
Move most docs to sphinx

Put documentation in a single website with Sphinx rather than having
information spread across markdown files that are implicitly organized
by the repo directory structure.

roshandathathri commented on docs/contributing/building.rst:
```diff
@@ -0,0 +1,221 @@
+   BUILD_DIR=$(pwd)/build
+   mkdir -p $BUILD_DIR
+   cd $BUILD_DIR
+   conan install $SRC_DIR/config --build=missing
```
nit: Add `SRC_DIR=$(pwd)`

roshandathathri commented on docs/contributing/building.rst:
```diff
@@ -0,0 +1,221 @@
+   mkdir -p $BUILD_DIR
+   cd $BUILD_DIR
+   conan install $SRC_DIR/config --build=missing
+   cmake -S $SRC_DIR -B $BUILD_DIR -DCMAKE_TOOLCHAIN_FILE=conan_paths.cmake -DKATANA_LANG_BINDINGS=python
```
Why are the python language bindings required? I haven't used it. 

I use these: ` -DKATANA_AUTO_CONAN=ON -D_GLIBCXX_USE_CXX11_ABI=ON`. Aren't these required?

ddn0 replied:
Going forward, we are going to assume people are going to use the Python language bindings. The build obviously supports other configuration options, but if you want to do something different, it is up to you to know what that means and that it is different from the "standard" build.

roshandathathri replied:
Then why isn't Python language bindings the default? 

What about the other two options I mentioned above? Are they not required?

ddn0 replied:
> Then why isn't Python language bindings the default?

Because this is a PR about the documentation of the current state.

> What about the other two options I mentioned above? Are they not required?

They have never been required.

roshandathathri commented on docs/contributing/testing.rst:
```diff
@@ -0,0 +1,239 @@
+``external/katana/python/katana/exmaple_utils.py``.  ``vN`` is a monotonically
+increasing version number. You can use a command ``inputs/update_inputs.sh`` to
+create create a new input collection. After creating the tar file, you will
+need to upload the file to the public S3 bucket.
```
I'm assuming "public" means that external contributor can also access and update it? Do we need to keep track of that?

ddn0 replied:
public just means globally accessible here. An external contributor won't be able update this directly.

roshandathathri replied:
Can you please add that in the docs because the above text talks about internal and external contributors?

ddn0 replied:
I don't understand. There is no mention of internal and external contributors in the `testing.rst` document.

I also don't see the point of saying "upload the file here; if you don't have authorization to upload, you won't be able to upload"

roshandathathri replied:
But it was mentioned in some other document. 

As this is an open repo, it would be good to be explicit if some statements only apply to internal contributors, especially because "public" can be misleading.

ddn0 replied:
I actually don't think it is good to draw this distinction. "contributor" is only used two times in the existing documentation, and it is mainly in reference to how PRs should be merged. In a situation where you have contributors with write access to this repo and those who do not, there is a question about the convention on how to merge changes for people who have write access. The ones who do not have write access must have a committer (someone with write access) merge their changes.

The people who have write access to some random S3 bucket is not related to (write) repo access at all. And again, it will be very obvious that you don't have access to this bucket because you cannot write to it.

roshandathathri replied:
Ok, that makes sense.

roshandathathri commented on libsupport/include/katana/Result.h:
```diff
@@ -16,6 +16,120 @@
+///
+///     Result<void> MakeList() {
+///       ...
+///       KATANA_CHECKED_CONTEXT(CheckNumber(n), "making number", n);
```
There's no `{}`?

roshandathathri reviewed (approved):
Thanks for writing this up!
